File: app/src/main/java/fernandagallina/weatherapp/WeatherFragment.java
package fernandagallina.weatherapp;

import android.content.res.Resources;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.os.Bundle;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.AppCompatActivity;
import android.text.format.DateFormat;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.TextView;
import android.widget.Toast;

import org.json.JSONObject;


import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;

/**
 * A placeholder fragment containing a simple view.
 */
public class WeatherFragment extends Fragment {
    Typeface weatherFont;
    TextView cityField;
    TextView detailsField;
    TextView currentTemperatureField;
    TextView weatherIcon;
    TextView timeField;
    TextView sunsetField;
    TextView sunriseField;

    View rootView;

    Handler handler;
    Handler handlerTime;

    String lat, lon, dt;
    long sunsetValue, sunriseValue;

    public WeatherFragment(){
        handler = new Handler(Looper.getMainLooper());
        handlerTime = new Handler(Looper.getMainLooper());
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        weatherFont = Typeface.createFromAsset(getActivity().getAssets(), "weather.ttf");
        updateWeatherData(new CityPreference(getActivity()).getCity());
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_main, container, false);
        cityField = (TextView)rootView.findViewById(R.id.city_field);
        detailsField = (TextView)rootView.findViewById(R.id.details_field);
        currentTemperatureField = (TextView)rootView.findViewById(R.id.current_temperature_field);
        weatherIcon = (TextView)rootView.findViewById(R.id.weather_icon);
        timeField = (TextView)rootView.findViewById(R.id.time_field);
        sunriseField = (TextView)rootView.findViewById(R.id.sunrise);
        sunsetField = (TextView)rootView.findViewById(R.id.sunset);

        weatherIcon.setTypeface(weatherFont);
        return rootView;
    }

    private void updateWeatherData(final String city){
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                final JSONObject json = RemoteFetch.getJSON(getActivity(), city);
                if(json == null){
                    handler.post(new Runnable(){
                        public void run(){
                            Toast.makeText(getActivity(),
                                    getActivity().getString(R.string.place_not_found),
                                    Toast.LENGTH_LONG).show();
                        }
                    });
                } else {
                    handler.post(new Runnable(){
                        public void run(){
                            renderWeather(json);
                        }
                    });
                }
            }
        };

        new Thread(runnable).start();
    }


    private void renderWeather(JSONObject json){
        try {

            cityField.setText(json.getString("name").toUpperCase(Locale.US));
//            +
//                    ", " +
//                    json.getJSONObject("sys").getString("country"));

            JSONObject details = json.getJSONArray("weather").getJSONObject(0);
            JSONObject main = json.getJSONObject("main");
            JSONObject coord = json.getJSONObject("coord");

            detailsField.setText(
                    details.getString("description").toUpperCase(Locale.US) +
                            "\n" + "Humidity: " + main.getString("humidity") + "%" +
                            "\n" + "Pressure: " + main.getString("pressure") + " hPa");

            currentTemperatureField.setText(
                    String.format("%.2f", main.getDouble("temp"))+ " ℃");

            dt = ""+ json.getLong("dt");
            lat = "" + coord.getDouble("lat");
            lon = "" + coord.getDouble("lon");
            sunriseValue = json.getJSONObject("sys").getLong("sunrise") * 1000;
            sunsetValue = json.getJSONObject("sys").getLong("sunset") * 1000;

            setWeatherIcon(details.getInt("id"), sunriseValue, sunsetValue);

            updateCityData();

        }catch(Exception e){
            Log.e("SimpleWeather", "One or more fields not found in the JSON data");
        }
    }

    private void setWeatherIcon(int actualId, long sunrise, long sunset){

        int id = actualId / 100;
        String icon = "";
        int color = R.color.black;
        Resources res = getResources();
        if(actualId == 800){
            long currentTime = new Date().getTime();
            if(currentTime>=sunrise && currentTime<sunset) {
                icon = getActivity().getString(R.string.weather_sunny);
                color = R.color.yellow;
            } else {
                icon = getActivity().getString(R.string.weather_clear_night);
                color = R.color.white;
            }
        } else {
            switch(id) {
                case 2 : icon = getActivity().getString(R.string.weather_thunder);
                    color = R.color.gray;
                    break;
                case 3 : icon = getActivity().getString(R.string.weather_drizzle);
                    break;
                case 5 : icon = getActivity().getString(R.string.weather_rainy);
                    break;
                case 6 : icon = getActivity().getString(R.string.weather_snowy);
                         color = R.color.white;
                    break;
                case 7 : icon = getActivity().getString(R.string.weather_foggy);
                         color = R.color.gray;
                    break;
                case 8 : icon = getActivity().getString(R.string.weather_cloudy);
                         color = R.color.gray;
                    break;
            }
        }
        weatherIcon.setText(icon);
        weatherIcon.setTextColor(getResources().getColor(color));
    }

    private void updateCityData() {
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
            final JSONObject json = RemoteFetch.getJSONGoogleMaps(lat, lon, dt);
            if(json == null){
                handlerTime.post(new Runnable(){
                    public void run(){
                        Toast.makeText(getActivity(),
                                getActivity().getString(R.string.datetime_not_found),
                                Toast.LENGTH_LONG).show();
                    }
                });
            } else {
                while (!Thread.currentThread().isInterrupted()) {
                    try {
                        handlerTime.post(new Runnable(){
                            @Override
                            public void run(){
                                renderTime(json);
                            }
                        });
                        Thread.sleep(10000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
            }
        };

        new Thread(runnable).start();
    }


    private void renderTime(JSONObject json){
        try {
            String status = json.getString("status");
            if(status.equals("OK")) {
                String timeZoneId = json.getString("timeZoneId");

                TimeZone tz = TimeZone.getTimeZone(timeZoneId);
                SimpleDateFormat destFormat = new SimpleDateFormat("HH:mm");
                destFormat.setTimeZone(tz);
                Date today = new Date();
                String s = destFormat.format(today);
                String ss = destFormat.format(new Date(sunsetValue));
                String sr = destFormat.format(new Date(sunriseValue));
                timeField.setText(s);
                sunriseField.setText(sr);
                sunsetField.setText(ss);
            }
        }catch(Exception e){
            Log.e("Google TimeZone API", "One or more fields not found in the JSON data");
        }
    }

    public void changeCity(String city){
        updateWeatherData(city);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        rootView = null; // cleaning up!
    }
}

File: app/src/main/kotlin/fernandagallina/weatherapp/weather/WeatherFragment.kt
package fernandagallina.weatherapp.weather

import android.graphics.Typeface
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.app.AlertDialog
import android.text.InputType
import android.view.*
import android.widget.EditText
import android.widget.Toast
import fernandagallina.weatherapp.App
import fernandagallina.weatherapp.CityPreference
import fernandagallina.weatherapp.R
import fernandagallina.weatherapp.di.DaggerWeatherComponent
import fernandagallina.weatherapp.di.WeatherModule
import fernandagallina.weatherapp.model.WeatherInfo
import kotlinx.android.synthetic.main.fragment_main.*
import java.text.SimpleDateFormat
import java.util.*
import javax.inject.Inject


class WeatherFragment : Fragment(), WeatherContract.View {

    @Inject
    lateinit var presenter: WeatherPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        getComponent()
        presenter.loadWeatherData(CityPreference(activity).city)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater!!.inflate(R.layout.fragment_main, container, false)
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater!!.inflate(R.menu.menu_main, menu)

        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item!!.itemId) {
            R.id.change_city -> {
                showInputDialog()
                return true
            }

            R.id.celsius_fahr -> {
                changeUnit()
                return true
            }

            else -> {
                return super.onOptionsItemSelected(item)
            }
        }
    }

    override fun setWeatherIcon(weatherInfo: WeatherInfo) {
       weatherIconTextView.typeface = Typeface.createFromAsset(activity.assets, "weather.ttf")

        val id = weatherInfo.weather[0].id / 100
        var icon = ""
        var color = R.color.black
        if (id == 8) {
            val currentTime = Date().time
            if (currentTime in weatherInfo.sys.sunrise..(weatherInfo.sys.sunset - 1)) {
                icon = activity.getString(R.string.weather_sunny)
                color = R.color.yellow
            } else {
                icon = activity.getString(R.string.weather_clear_night)
                color = R.color.black
            }
        } else {
            when (id) {
                2 -> {
                    icon = activity.getString(R.string.weather_thunder)
                    color = R.color.gray
                }
                3 -> icon = activity.getString(R.string.weather_drizzle)
                5 -> icon = activity.getString(R.string.weather_rainy)
                6 -> {
                    icon = activity.getString(R.string.weather_snowy)
                    color = R.color.black
                }
                7 -> {
                    icon = activity.getString(R.string.weather_foggy)
                    color = R.color.gray
                }
                8 -> {
                    icon = activity.getString(R.string.weather_cloudy)
                    color = R.color.gray
                }
            }
        }
        weatherIconTextView.text = icon
        weatherIconTextView.setTextColor(resources.getColor(color))

        val destFormat = SimpleDateFormat("HH:mm", Locale("en", "US"))
        val today = Date()

        timeFieldTextView.text = destFormat.format(today)
        sunriseFieldTextView.text = destFormat.format(Date(weatherInfo.sys.sunrise * 1000))
        sunsetFieldTextView.text = destFormat.format(Date(weatherInfo.sys.sunset * 1000))
        currentTemperatureField.text = weatherInfo.main.temp.toString() + "°C"
        cityFieldTextView.text = weatherInfo.name.toUpperCase(Locale.ENGLISH) + ", " + weatherInfo.sys.country
        detailsFieldTextView.text = weatherInfo.weather[0].description + "\n" +
                "Humidity: " + weatherInfo.main.humidity.toString() + "% \n" +
                "Pressure: " + weatherInfo.main.pressure.toString() + " hPa"
    }

    override fun showErrorMessage(localizedMessage: String?) {
        Toast.makeText(context, localizedMessage, Toast.LENGTH_LONG).show()
    }

    private fun getComponent() {
        DaggerWeatherComponent.builder()
                .netComponent((activity.applicationContext as App).netComponent)
                .weatherModule(WeatherModule(this))
                .build().inject(this)
    }

    private fun showInputDialog() {
        val builder = AlertDialog.Builder(context)
        builder.setTitle("Change city")
        val input = EditText(context)
        input.inputType = InputType.TYPE_CLASS_TEXT
        builder.setView(input)
        builder.setPositiveButton("Search") { _, _ -> changeCity(input.text.toString()) }
        builder.show()
    }

    fun changeCity(city: String) {
        presenter.loadWeatherData(city)
    }

    private fun changeUnit() {
        // todo
    }
}
